pages/tutorials/3.0/window/events.md

```diff
@@ -7,259 +7,329 @@
     }
+
+    // Remainder of main loop
 }
 ```
 
-Read the above paragraph once again and make sure that you fully understand it, the [`sf::Event`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Event.php "sf::Event documentation") union is the cause of many problems for inexperienced programmers.
+Note how the API for getting events has changed slightly.
+`sf::WindowBase::pollEvent` and `sf::WindowBase::waitEvent` now return a `std::optional<sf::Event>`.
+These two functions _might_ return an event but they might not.
+C++ lets you deduce the template parameter which is why you can write `const std::optional event` instead of `const std::optional<sf::Event> event`.
+`const auto event` is another valid choice if you prefer a more terse expression.
+
+### `sf::WindowBase::handleEvents`
 
-Alright, now we can see what events SFML supports, what they mean and how to use them properly.
+The second option for processing events is via the `sf::WindowBase::handleEvents` function.
+This functions performs event visitation.
+What this means is that you can provide callbacks which take different event subtypes as arguments.
+Alternatively you may provide an object (or objects) with `operator()` implementations which handle the event subtypes you want to process.
+Notably you do not have to provide callbacks for all possible event subtypes.
+Depending on the current active event subtype, the corresponding callback is called.
+Here's what that looks like:
+
+```cpp
+const auto onClose = [&window](const sf::Event::Closed&)
+{
+    window.close();
+};
+
+const auto onKeyPressed = [&window](const sf::Event::KeyPresssed& keyPressed)
+{
+    if (keyPressed.scancode == sf::Keyboard::Scancode::Escape)
+        window.close();
+};
+
+while (window.isOpen())
+{
+    window.handleEvents(onClose, onKeyPressed);
+
+    // Remainder of main loop
+}
+```
 
 ## The Closed event
 
-The `sf::Event::Closed` event is triggered when the user wants to close the window, through any of the possible methods the window manager provides ("close" button, keyboard shortcut, etc.). This event only represents a close request, the window is not yet closed when the event is received.
+The `sf::Event::Closed` event is triggered when the user wants to close the window through any of the possible methods the window manager provides ("close" button, keyboard shortcut, etc.).
+This event only represents a close request.
+The window is not yet closed when the event is received.
 
-Typical code will just call `window.close()` in reaction to this event, to actually close the window. However, you may also want to do something else first, like saving the current application state or asking the user what to do. If you don't do anything, the window remains open.
+Typically, one will just call `window.close()` in reaction to this event to actually close the window.
+However, you may also want to do something else first like saving the current application state or asking the user what to do.
```

Comment:
I trust your comma skills, but to me there's a clear pause here... 😅 

```suggestion
However, you may also want to do something else first, like saving the current application state or asking the user what to do.
```

Reply:
Commas tend to represent a pause in one's speech but they're not defined as such. In this case I don't think the comma is correct. It may exist in a grey area where it's not necessarily correct nor incorrect to use it. English is a bit of a fuzzy language.

Reply:
Grammarly has a special rule for "such as", which one could use in place of "like".

![image](https://github.com/user-attachments/assets/2e965a37-48bb-4a9e-a18a-6286efdd6e3e)

Given that this a nonrestrictive clause, it would require a comma.

> English is a bit of a fuzzy language.

From a German perspective, that's a bit of an understatement. 😄  
The German language has a central and detailed grammatical definition of basically everything. Where English has countless different definitions.